src/overlays/actors/ovl_Bg_Mori_Rakkatenjo/z_bg_mori_rakkatenjo.c

```diff
@@ -20,35 +42,199 @@ const ActorInit Bg_Mori_Rakkatenjo_InitVars = {
+            if (this->timer <= 0) {
+                BgMoriRakkatenjo_SetupFall(this);
+            }
+        } else {
+            this->timer = 100;
+        }
+    }
+    if (this->timer < 20) {
+        func_800788CC(NA_SE_EV_BLOCKSINK - SFX_FLAG);
+    }
+}
+
+void BgMoriRakkatenjo_SetupFall(BgMoriRakkatenjo* this) {
+    this->actionFunc = BgMoriRakkatenjo_Fall;
+    this->bounceCount = 0;
+    this->dyna.actor.velocity.y = 0.0f;
+}
+
+void BgMoriRakkatenjo_Fall(BgMoriRakkatenjo* this, GlobalContext* globalCtx) {
+    static f32 bounceVel[] = { 4.0f, 1.5f, 0.4f, 0.1f };
+    s32 temp;
+    Actor* thisx = &this->dyna.actor;
+    s32 quake;
+
+    Actor_MoveForward(thisx);
+    if ((thisx->velocity.y < 0.0f) && (thisx->posRot.pos.y <= 403.0f)) {
+        if (this->bounceCount >= ARRAY_COUNT(bounceVel)) {
+            BgMoriRakkatenjo_SetupRest(this);
+        } else {
+            if (this->bounceCount == 0) {
+                this->fallCount++;
+                func_800788CC(NA_SE_EV_STONE_BOUND);
+                func_800AA000(SQ(thisx->yDistFromLink), 0xFF, 0x14, 0x96);
+            }
+            thisx->posRot.pos.y =
+                403.0f - (thisx->posRot.pos.y - 403.0f) * bounceVel[this->bounceCount] / fabsf(thisx->velocity.y);
+            thisx->velocity.y = bounceVel[this->bounceCount];
+            this->bounceCount++;
+            quake = Quake_Add(ACTIVE_CAM, 3);
+            Quake_SetSpeed(quake, -0x3CB0);
+            Quake_SetQuakeValues(quake, 5, 0, 0, 0);
+            Quake_SetCountdown(quake, 5);
+        }
+    }
+}
+
+void BgMoriRakkatenjo_SetupRest(BgMoriRakkatenjo* this) {
+    this->actionFunc = BgMoriRakkatenjo_Rest;
+    this->dyna.actor.posRot.pos.y = 403.0f;
+    this->timer = 20;
+}
+
+void BgMoriRakkatenjo_Rest(BgMoriRakkatenjo* this, GlobalContext* globalCtx) {
+    if (this->timer <= 0) {
+        BgMoriRakkatenjo_SetupRise(this);
+    }
+}
+
+void BgMoriRakkatenjo_SetupRise(BgMoriRakkatenjo* this) {
+    this->actionFunc = BgMoriRakkatenjo_Rise;
+    this->dyna.actor.velocity.y = -0.1f;
+}
+
+void BgMoriRakkatenjo_Rise(BgMoriRakkatenjo* this, GlobalContext* globalCtx) {
+    Math_SmoothScaleMaxMinF(&this->dyna.actor.velocity.y, 5.0f, 0.06f, 0.1f, 0.0f);
+    this->dyna.actor.posRot.pos.y += this->dyna.actor.velocity.y;
+    if (this->dyna.actor.posRot.pos.y >= 683.0f) {
+        BgMoriRakkatenjo_SetupWait(this);
+    }
+}
+
+void BgMoriRakkatenjo_Update(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    BgMoriRakkatenjo* this = THIS;
+
+    if (this->timer > 0) {
+        this->timer--;
+    }
+    this->actionFunc(this, globalCtx);
+    if (BgMoriRakkatenjo_LinkUnder(this, globalCtx)) {
+        if (sCamSetting == 0) {
+            osSyncPrintf("camera changed (mori rakka tenjyo) ... \n");
+            sCamSetting = globalCtx->cameraPtrs[0]->setting;
+            Camera_SetCameraData(globalCtx->cameraPtrs[0], 1, &this->dyna.actor, NULL, 0, 0, 0);
```

Comment:
huh... the third argument to `Camera_SetCameraData` should be a `CutsceneCameraPoint*`, how does this even work?

Reply:
Camera_SetCameraData's second argument is bitflags for which data to set. Since only the first bitflag is set, it's ignoring that pointer entirely.

Reply:
`Camera_SetCameraData` actually takes a `void*` according to the z_camera.c PR

Reply:
Hm, I wonder why krim changed that because the only data0 ( https://github.com/krimtonz/oot/blob/z_camera/src/code/z_camera.c#L8086 ) usage I can see casts it to `CutsceneCameraPoint*` https://github.com/krimtonz/oot/blob/z_camera/src/code/z_camera.c#L5522

Reply:
Krim's answer

> The fact that data1 get used for other things and it didn't make sense to me to keep it at a fixed type when it's paired usage is also not a fixed type

now where does this kind of data0 data get used, wait and see